# Week 1–2: Data Cleaning and Exploratory Data Analysis (EDA)

## AnalystLab Africa Data Analytics Internship – Batch C

**Intern:** Asres Yelia

---

# Project Overview

This notebook documents the data cleaning and exploratory data analysis (EDA) process for the **Online Retail Dataset** as part of Week 1–2 of the AnalystLab Africa Data Analytics Internship.

The objective is to assess data quality, clean the dataset, explore purchasing patterns, and generate meaningful business insights using Python and Pandas.

The cleaned dataset will serve as a reliable foundation for further analysis and visualization.

# Business Understanding

Understanding the business context is an important first step in the data analytics process. Before analyzing the data, it is necessary to understand what the dataset represents and what business questions the analysis aims to answer.

The Online Retail Dataset contains transactional data from a UK-based online retail company that sells unique gift products to customers across different countries.

The company records information about customer purchases, including products sold, quantities, prices, transaction dates, customer details, and locations.

The purpose of this analysis is to examine the quality of the transaction data, prepare it for analysis, and discover patterns that can help understand sales performance and customer purchasing behavior.

The analysis will focus on answering key business questions such as:

- Which products generate the highest sales?
- Which countries contribute the most revenue?
- What purchasing patterns can be identified?
- How can customer transaction data be improved for better decision-making?

By cleaning and analyzing this dataset, we aim to transform raw transaction records into reliable information that can support business decisions related to sales performance, customer understanding, and operational improvements.

---

# Project Objectives

The objectives of this project are to:

- Understand the structure of each dataset.
- Assess data quality issues.
- Identify missing values and duplicate records.
- Correct inconsistent data where necessary.
- Prepare clean datasets for analysis.
- Perform exploratory data analysis (EDA).
- Create informative visualizations.
- Generate meaningful business insights.
- Document the entire analytical process professionally.

---


# Dataset Description

The dataset used in this project is the **Online Retail Dataset**.

It contains transactional records from a UK-based online retail company between 2010 and 2011. Each record represents a product-level transaction made by a customer.

The dataset contains information about sales transactions, products, customers, and geographical locations.

## Dataset Features

| Column | Description |
|---|---|
| InvoiceNo | Unique identifier for each transaction |
| StockCode | Unique code assigned to each product |
| Description | Name or description of the product |
| Quantity | Number of products purchased |
| InvoiceDate | Date and time when the transaction occurred |
| UnitPrice | Price per unit of product |
| CustomerID | Unique identifier for each customer |
| Country | Customer's country location |

The dataset will be analyzed to understand data quality issues, customer purchasing behavior, and sales patterns.

# Tools and Libraries

This project uses Python-based data analytics tools to clean, analyze, and explore the Online Retail Dataset.

## Programming Language

- Python

## Libraries

- Pandas: Used for data loading, cleaning, and data manipulation.
- NumPy: Used for numerical operations.
- Matplotlib: Used for data visualization.
- Seaborn: Used for statistical visualizations.

## Development Environment

- Jupyter Notebook
- VS Code

## Version Control

- Git & GitHub

These tools support a reproducible and professional data analytics workflow.

In [2]:
# import libraries
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

# Loading the Dataset

The next step is to load the Online Retail dataset into a Pandas DataFrame.

A DataFrame is a two-dimensional table structure that allows us to inspect, clean, and analyze the data.

The dataset is loaded using `pd.read_csv()`.

The encoding parameter is set to `latin1` because the dataset contains special characters that may not be supported by the default UTF-8 encoding.

In [4]:
# load the dataset
online = pd.read_csv("../data/raw/OnlineRetail.csv", encoding="latin1")

# Confirm Dataset Loading

After loading the dataset, the next step is to verify that the data has been successfully imported.

The `head()` function displays the first five rows of the dataset and helps us understand the initial structure of the data.

In [6]:
# View the first few rows of the dataset
online.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


In [8]:

# View the dimensions of the dataset
online.shape


(541909, 8)

In [10]:
# View information about the dataset
online.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


## Previewing the Data (.head())

The `.head()` function shows the first 5 rows of the dataset.

This helps me:
- Understand column structure
- See sample values
- Identify data types visually

In [ ]:
online.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


## Observations from `.head()`

From the preview, I observe:

- Each row represents a single product purchase
- One invoice can contain multiple rows (multiple products per order)
- The dataset includes product, customer, and country information
- Transaction time is stored in `InvoiceDate`
- Customer behavior analysis will be possible using `CustomerID`

In [ ]:
online.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


##  Data Quality Assessment

From the `.info()` output, I observed:

###  Missing Values:
- `Description` has some missing values
- `CustomerID` has a significant number of missing values

###  Data Types:
- `InvoiceDate` is stored as object (needs conversion to datetime)
- `CustomerID` is float but should be converted to integer or string

###  Dataset Size:
- 541,909 total records
- 8 columns in total

##  Business Interpretation

- Missing `CustomerID` suggests some transactions are not linked to registered customers
- This may represent guest checkout or incomplete records
- Converting `InvoiceDate` is important for time-based analysis
- Data cleaning will be required before analysis

In [ ]:
# Missing values in the dataset
online.isnull().sum()



InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [ ]:
# duplicate rows in the dataset
online.duplicated().sum()

np.int64(5268)

##  Data Quality Issues Identified

###  Missing Values:
- `Description`: 1,454 missing values
- `CustomerID`: 135,080 missing values (significant issue)

###  Duplicate Records:
- 5,268 duplicate rows detected


## Interpretation

- Missing `CustomerID` may indicate guest transactions or incomplete customer tracking
- Missing product descriptions may affect product-level analysis
- Duplicate records must be removed to ensure accurate analysis results

## Removing Duplicate Records

Duplicate rows can distort analysis results such as:
- Total sales
- Product popularity
- Customer behavior

Therefore, I will remove duplicate entries from the dataset.

In [ ]:
# Remove duplicate rows
online = online.drop_duplicates()

In [ ]:
# checking again for duplicate rows
online.duplicated().sum()

np.int64(0)

##  Duplicate Check After Cleaning

After removing duplicates, I re-checked the dataset.

Result:
- No duplicate rows found (0 duplicates)

This confirms that the data cleaning step was successful.

## Handling Missing Values

The dataset contains missing values in:
- Description
- CustomerID

These must be handled before analysis.

In [ ]:
# Remove rows with missing values containing CustomerID
online = online.dropna(subset=["CustomerID"])


In [ ]:
# Remove rows with missing values

online = online.dropna(subset=["Description"])
# print the number of rows and columns in the dataset
print("Number of rows:", online.col

Number of rows: 401604
Number of columns: 8
